### PR TITLE
Vsts reporter

### PR DESCRIPTION
**Pull request by onovotny**
This is the implementation for #1627.

Note it doesn't work yet since getting the access token in YAML builds doesn't appear to be complete yet:
https://github.com/Microsoft/vsts-agent/blob/master/docs/preview/yamlgettingstarted-token.md

Pending verification once the access token is available.

To test: create a YAML based build of xunit in a VSTS instance and point to the `.vsts.ci.yml` file as the yaml file. It should build and you should see test results in the tests area of the build results.

### CODE REVIEW

File: src/xunit.runner.reporters/VstsClient.cs [hunk 1]
```diff
@@ -0,0 +1,310 @@
...
+        {
+            int? runId = null;
+            try
+            {
+                runId = await CreateTestRun();
+
+                while (!shouldExit || !addQueue.IsEmpty || !updateQueue.IsEmpty)
+                {
+                    workEvent.Wait();   // Wait for work
+                    workEvent.Reset();  // Reset first to ensure any subsequent modification sets
```
Comment (bradwilson) [inline src/xunit.runner.reporters/VstsClient.cs]: This is a race condition. If you want this behavior, switch from `ManualResetEventSlim` to `AutoResetEventSlim`.

File: src/xunit.runner.reporters/VstsClient.cs [hunk 2]
```diff
@@ -0,0 +1,310 @@
...
+
+                    if (previousErrors)
+                        break;
+
+                    // We have to do add's before update because we need the test id from the add to inject into the update
+                    await SendTestResults(true, runId.Value, aq.ToArray()).ConfigureAwait(false);
+                    await SendTestResults(false, runId.Value, uq.ToArray()).ConfigureAwait(false);
+                }
+            }
+            catch { }
```
Comment (bradwilson) [inline src/xunit.runner.reporters/VstsClient.cs]: This should log a diagnostic message, at a minimum.

File: src/xunit.runner.reporters/VstsClient.cs [hunk 3]
```diff
@@ -0,0 +1,310 @@
...
+            finally
+            {
+                try
+                {
+                    if (runId.HasValue)
+                        await FinishTestRun(runId.Value);
+                }
+                catch
```
Comment (bradwilson) [inline src/xunit.runner.reporters/VstsClient.cs]: Diagnostic message

File: src/xunit.runner.reporters/VstsReporter.cs [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+            => "vsts";
+
+        public IMessageSink CreateMessageHandler(IRunnerLogger logger)
+        {
+            var accessToken = EnvironmentHelper.GetEnvironmentVariable("VSTS_ACCESS_TOKEN");
+            var collectionUri = EnvironmentHelper.GetEnvironmentVariable("SYSTEM_TEAMFOUNDATIONCOLLECTIONURI");
+            var teamProject = EnvironmentHelper.GetEnvironmentVariable("SYSTEM_TEAMPROJECT");
+
+            //build id is the id associated with the build number, which we will use to associate the test run with
+            var buildId = Convert.ToInt32(EnvironmentHelper.GetEnvironmentVariable("BUILD_BUILDID"));
```
Comment (bradwilson) [inline src/xunit.runner.reporters/VstsReporter.cs]: Seems like all these environment variables are needed, so shouldn't they all be detected for above?

File: src/xunit.runner.reporters/VstsClient.cs (hunk 1 above)
Comment (onovotny) [reply src/xunit.runner.reporters/VstsClient.cs]: We use the same pattern in `AppVeyorClient`, I assume it needs updating there too?

File: src/xunit.runner.reporters/VstsReporter.cs (hunk 4 above)
Comment (onovotny) [reply src/xunit.runner.reporters/VstsReporter.cs]: Those are all defined as being present on the build agent by: https://docs.microsoft.com/en-us/vsts/build-release/concepts/definitions/build/variables?tabs=batch

Question is if it's worth checking all of them? The important one is to detect we are running on the agent and that the access token is properly set.

I'm not sure about the name, should we specify `VSTS_ACCESS_TOKEN` or something more specific like `XUNIT_ACCESS_TOKEN`?

**Comment by onovotny** [comment]
Confirmed working now!

image

**Pull request merged**